hibernate-core/src/main/java/org/hibernate/metamodel/internal/MetadataContext.java

```diff
@@ -421,19 +421,13 @@ private void applyIdMetadata(PersistentClass persistentClass, IdentifiableDomain
 		if ( persistentClass.hasIdentifierProperty() ) {
 			final Property declaredIdentifierProperty = persistentClass.getDeclaredIdentifierProperty();
 			if ( declaredIdentifierProperty != null ) {
-				final SingularPersistentAttribute<?, Object> idAttribute = attributeFactory.buildIdAttribute(
-						identifiableType,
-						declaredIdentifierProperty
-				);
-
-				//noinspection unchecked rawtypes
-				( ( AttributeContainer) identifiableType ).getInFlightAccess().applyIdAttribute( idAttribute );
+				applyIdAttribute( identifiableType, declaredIdentifierProperty );
 			}
 			else if ( persistentClass.getIdentifier() instanceof Component
 					&& persistentClass.getIdentifierProperty() != getSuperclassIdentifier( persistentClass ) ) {
-				// If the identifier is a generic component, we have to call buildIdAttribute anyway,
-				// as this will create and register the EmbeddableType for the subtype
-				attributeFactory.buildIdAttribute( identifiableType, persistentClass.getIdentifierProperty() );
+				// If the identifier is a different component than the superclass, we have to apply the id
+				// attribute anyway, as this will create and register the EmbeddableType for the subtype
+				applyIdAttribute( identifiableType, persistentClass.getIdentifierProperty() );
```

Comment:
Unfortunately, this is not strictly correct. With this change, the metamodel types now have attribute objects for `id` that are different. This is a problem for the static metamodel and simply not how JPA "ticks".

I just had a recent discussion with @Sanne about a very similar example. IMO if we want to support this, then we have to implement this in `EmbeddedSqmPathSource` and `SqmEmbeddedValuedSimplePath`, by capturing the type variable assignments on the path and lookup the correct `EmbeddedSqmPathSource` based on that when finding a sub path source.

Reply:
> by capturing the type variable assignments on the path and lookup the correct EmbeddedSqmPathSource based on that when finding a sub path source

I'd like to try that, but I guess the correct `EmbeddedSqmPathSource`s with type parameters must be somehow *registered* to be able to retrieve them later. Any pointers / example on how it could be done?

Reply:
I updated the PR with a new solution that registers *concrete* generic embeddable attributes for identifier properties and retrieves them when creating `SqmPath`s.

I have opened another issue [HHH-16238](https://hibernate.atlassian.net/browse/HHH-16238) which expands on this for the handling of multiple non-identifier generic embeddable properties.